### PR TITLE
IGNITE-19204 Implement ODBC connection establishment

### PR DESCRIPTION
**Pull request by isapego**
- Ported network part from Ignite 2;
- Implemented connection establishment;
- Added test for connection establishment.

### CODE REVIEW

File: modules/platforms/cpp/ignite/client/ignite_client_configuration.h [hunk 1]
```diff
@@ -77,6 +85,26 @@ class ignite_client_configuration {
         ignite_client_configuration::m_endpoints.assign(endpoints.begin(), endpoints.end());
     }
 
+    /**
+     * Set endpoints.
+     *
+     * Examples of supported formats:
+     * - 192.168.1.25 - Default port is used, see DEFAULT_PORT;
+     * - 192.168.1.25:780 - Custom port;
+     * - 192.168.1.25:780..787  - Custom port range - ports are checked from
```
Comment (ptupitsyn) [inline modules/platforms/cpp/ignite/client/ignite_client_configuration.h:94]: We agreed to remove port ranges (see https://issues.apache.org/jira/browse/IGNITE-19601) - should we get rid of them right here?

File: modules/platforms/cpp/ignite/network/detail/linux/tcp_socket_client.h [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+            std::stringstream converter;
+            converter << port;
+            std::string strPort = converter.str();
+
+            // Resolve the server address and port
+            addrinfo *result{NULL}; // NOLINT(modernize-use-nullptr)
+            int res = getaddrinfo(hostname, strPort.c_str(), &hints, &result);
+
+            if (res != 0)
```
Comment (ptupitsyn) [inline modules/platforms/cpp/ignite/network/detail/linux/tcp_socket_client.h:88]: Is the `res` useful, should we include it in the error message?

File: modules/platforms/cpp/ignite/network/detail/win/sockets.cpp [hunk 3]
```diff
@@ -82,10 +83,10 @@ void try_set_socket_options(SOCKET socket, int buf_size, BOOL no_delay, BOOL out
 
     int res = setsockopt(socket, SOL_SOCKET, SO_KEEPALIVE, reinterpret_cast<char *>(&keep_alive), sizeof(keep_alive));
 
-    // TODO: IGNITE-17606 Disable keep-alive once heartbeats are implemented.
+    // TODO: Disable keep-alive once heartbeats are implemented.
```
Comment (ptupitsyn) [inline modules/platforms/cpp/ignite/network/detail/win/sockets.cpp:86]: Is there a ticket?

File: modules/platforms/cpp/ignite/network/detail/win/tcp_socket_client.h [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+
+        std::stringstream converter;
+        converter << port;
+        std::string strPort = converter.str();
+
+        // Resolve the server address and port
+        addrinfo *result{NULL}; // NOLINT(modernize-use-nullptr)
+        int res = getaddrinfo(hostname, strPort.c_str(), &hints, &result);
+
+        if (res != 0)
```
Comment (ptupitsyn) [inline modules/platforms/cpp/ignite/network/detail/win/tcp_socket_client.h:84]: Same as above - is `res` useful for the error message?

**Review by ptupitsyn — APPROVED** [review]

File: modules/platforms/cpp/ignite/client/ignite_client_configuration.h (hunk 1 above)
Comment (isapego) [reply modules/platforms/cpp/ignite/client/ignite_client_configuration.h:94]: I'll do that for ODBC, but I'd prefer to do it in a separate ticket for client. I remove port ranges from the comments though.

File: modules/platforms/cpp/ignite/network/detail/linux/tcp_socket_client.h (hunk 2 above)
Comment (isapego) [reply modules/platforms/cpp/ignite/network/detail/linux/tcp_socket_client.h:88]: Good catch. Fixed.

File: modules/platforms/cpp/ignite/network/detail/win/sockets.cpp (hunk 3 above)
Comment (isapego) [reply modules/platforms/cpp/ignite/network/detail/win/sockets.cpp:86]: Yeah, and the ticket was actually correct. Seems like I messed up something here. Reverted changes.

File: modules/platforms/cpp/ignite/network/detail/win/tcp_socket_client.h (hunk 4 above)
Comment (isapego) [reply modules/platforms/cpp/ignite/network/detail/win/tcp_socket_client.h:84]: Fixed.

**Review by ptupitsyn — APPROVED** [review]

**Pull request merged**